gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/upgrade/upgrader/IntegrationPrimaryOwnerUpgrader.java

```diff
@@ -0,0 +1,225 @@
+    private static final String DEFAULT_SOURCE = "system";
+
+    private final EnvironmentRepository environmentRepository;
+    private IntegrationRepository integrationRepository;
+    private final ApiRepository apiRepository;
+    private final MembershipRepository membershipRepository;
+    private final RoleRepository roleRepository;
+
+    @Autowired
+    public IntegrationPrimaryOwnerUpgrader(
+        @Lazy EnvironmentRepository environmentRepository,
+        @Lazy IntegrationRepository integrationRepository,
+        @Lazy ApiRepository apiRepository,
+        @Lazy MembershipRepository membershipRepository,
+        @Lazy RoleRepository roleRepository
+    ) {
+        this.environmentRepository = environmentRepository;
+        this.integrationRepository = integrationRepository;
+        this.apiRepository = apiRepository;
+        this.membershipRepository = membershipRepository;
+        this.roleRepository = roleRepository;
+    }
+
+    @Override
+    public boolean upgrade() {
+        try {
+            environmentRepository
+                .findAll()
+                .forEach(environment -> {
+                    ExecutionContext executionContext = new ExecutionContext(environment);
+                    try {
+                        setPrimaryOwnersForIntegrations(executionContext);
+                    } catch (TechnicalException e) {
+                        log.error("An error occurs while updating primary owner for integrations", e);
+                    }
+                });
+        } catch (Exception e) {
+            log.error("failed to apply {}", getClass().getSimpleName(), e);
+            return false;
+        }
+
+        return true;
+    }
+
+    private void setPrimaryOwnersForIntegrations(ExecutionContext executionContext) throws TechnicalException {
+        log.info("Start to set primary owners for integrations");
+
+        Pageable pageable = new PageableBuilder().pageNumber(0).pageSize(PAGE_SIZE).build();
+        Page<Integration> integrationPage = integrationRepository.findAllByEnvironment(executionContext.getEnvironmentId(), pageable);
+        var totalPageNumber = (int) Math.ceil((double) integrationPage.getTotalElements() / PAGE_SIZE);
+
+        setPrimaryOwnersForIntegrations(integrationPage, executionContext);
+
+        // Cover others pages if required
+        if (integrationPage.getTotalElements() > integrationPage.getPageElements()) {
+            for (int pageIdx = 1; pageIdx < totalPageNumber; pageIdx++) {
+                var nextIntegrationPageable = new PageableBuilder().pageNumber(pageIdx).pageSize(PAGE_SIZE).build();
+                var nextIntegrationPage = integrationRepository.findAllByEnvironment(
+                    executionContext.getEnvironmentId(),
+                    nextIntegrationPageable
+                );
+                setPrimaryOwnersForIntegrations(nextIntegrationPage, executionContext);
+            }
+        }
```

Comment:
To simplify and write the setPrimaryOwnersForIntegrations() call one time

```suggestion
        int handledIntegrations = 0;
        Page<Integration> integrationPage;
        do {
            currentPage = integrationRepository.findAllByEnvironment(executionContext.getEnvironmentId(), new PageableBuilder().pageNumber(handledIntegrations / PAGE_SIZE).pageSize(PAGE_SIZE).build());
            handledIntegrations += currentPage.getPageElements();

            setPrimaryOwnersForIntegrations(integrationPage, executionContext);
        } while (handledIntegrations < currentPage.getTotalElements());
```